Port to ROS 2

clalancette opened the pull request:
This PR is a port of joint_state_publisher to ROS 2.  I did the initial port, but got a bunch of help from people along the way (thanks @gonzodepedro, @schornakj, @emersonknapp, @sloretz, and @pbeeson).

Besides the port to ROS 2, this PR does a number of other things to improve the code:

1.  It adds a README describing the subscribed topics, published topics, and parameters.
1.  If the GUI is in use, and the GUI is closed, this now quits the program (which I think is a little more intuitive).
1.  Allows the joint_state_publisher to get its URDF either from the command-line, or from the `/robot_description` topic.  In the latter case, this node will actually do nothing until something is received on `/robot_description`, at which point it will start publishing.  If both the command-line option is specified and the `/robot_description` topic is available, the command-line option takes precedence.

I've done quite some testing on this to make sure all of the options work.  In my testing with a simple URDF, it does seem to work fairly well.  There are a few TODOs left:

1.  ~~Despite the fact that we `allow_undeclared_parameters`, there doesn't seem to be a good way to get at undeclared parameters in the rclpy API.  Thus, we can't really support `dependent_joints` and `zeros` until we figure that out.  That will be a PR for another time (maybe with some core rclpy changes as well).~~  I actually found a way to do this, so this should work.
1.  ~~While `rclpy.Rate` is available in Eloquent, I wanted to make sure this was backwards-compatible with Dashing as well.  So I left the TODO about switching to Rate in there.~~ This no longer applies since we switched to a timer.

This also fixes #20

sloretz commented:
> Despite the fact that we allow_undeclared_parameters, there doesn't seem to be a good way to get at undeclared parameters in the rclpy API. Thus, we can't really support dependent_joints and zeros until we figure that out. That will be a PR for another time (maybe with some core rclpy changes as well).

It looks like this can be made to work with `automatically_declare_parameters_from_overrides`. I'm having trouble understanding the meaning of the options to the parameters API, but it seems `automatically_declare_parameters_from_overrides` declares any parameter that is passed via yaml (Not just parameters via `parameter_overrides`), and `allow_undeclared_parameters` doesn't seem to be necessary.

```yaml
/**:
  ros__parameters:
    dependent_joints:
      joint_D: {parent: joint_A, factor: 3 }
      joint_E: {parent: joint_B }
      joint_F: {parent: joint_C, factor: -1 }
```

```python3
import rclpy
from rclpy.node import Node


if __name__ == '__main__':
    rclpy.init()
    node = Node('mock_jsp', automatically_declare_parameters_from_overrides=True)
    print(node.get_parameters_by_prefix('dependent_joints'))
    rclpy.shutdown()
```

```console
$ python3 mockjsp.py --ros-args --params-file ./jsp_params.yaml --
{'joint_D.parent': <rclpy.parameter.Parameter object at 0x7fd4103f9780>, 'joint_D.factor': <rclpy.parameter.Parameter object at 0x7fd4103f9be0>, 'joint_E.parent': <rclpy.parameter.Parameter object at 0x7fd4103a1390>, 'joint_F.parent': <rclpy.parameter.Parameter object at 0x7fd4156cdc88>, 'joint_F.factor': <rclpy.parameter.Parameter object at 0x7fd407c74160>}
```

clalancette commented:
@sloretz You are totally right, that works.  Thanks!  I've added back in support for dependent joints and zeros in 475e12f1d0f210fa28610434f9f3a9dbb5024ef0

clalancette commented:
FYI that this is probably on hold until we decide what to do about #31.  Once we've merged that, I'll likely rebase/rewrite this on top of that.

clalancette commented:
I've now redone this on top of #31 .  Still needs to wait until that is merged, but once that is in we can easily rebase and review this.

clalancette committed:
Port to ROS 2.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add in copyright and XML lint tests.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
All right, the ROS 1 stuff has been merged into noetic-devel.  I've updated the `ros2` branch to be the same as noetic-devel, and rebased this branch on top of that.  @sloretz This one should be ready for a review.

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -4,29 +4,27 @@ This contains a package for publishing `sensor_msgs/JointState` messages for a r
-It has been moved to this repo as described by [`ros/robot_model#195`](https://github.com/ros/robot_model/issues/195)
+Given a URDF (either passed on the command-line or via the `/robot_description` topic), this node
+will continually publish default values for all of the movable joints in the URDF to the `/joint_states` topic.
+In combination with `robot_state_publisher`, this can update the TF2 transforms for the robot on the fly.
```
While `robot_state_publisher` can `update TF2 transforms for the robot on the fly`, I'd describe `joint_state_publisher`'s contribution as making sure all transforms exist even when a joint doesn't have encoder data.

clalancette replied:
Yeah, I was trying to get that across with the previous sentence.  But maybe I muddled the waters too much by talking about `robot_state_publisher` here.  I think I'll just remove this sentence, and slightly change the wording in the previous sentence.

clalancette replied:
Fixed in 2917803

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -139,38 +139,116 @@ def init_urdf(self, robot):
+            pass
+
+    def __init__(self, urdf_file):
+        self.node = rclpy.create_node('joint_state_publisher', allow_undeclared_parameters=True,
```
I'd inherit from `rclpy.node.Node` so we could compose `joint_state_publisher` if we ever get around to making a component container for `rclpy`.

clalancette replied:
Good point.  Fixed in 2917803.

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -4,29 +4,27 @@ This contains a package for publishing `sensor_msgs/JointState` messages for a r
 * `source_list` (array of strings) - Each string in this array represents a topic name.  For each string, create a subscription to the named topic of type `sensor_msgs/JointStates`.  Publication to that topic will update the joints named in the message.  Defaults to an empty array.
-* `zeros` (dictionary of string -> float) - A dictionary of joint_names to initial starting values for the joint.  Defaults to an empty dictionary, in which case 0.0 is assumed as the zero for all joints.
-* `dependent_joints` (dictionary of string -> dictionary of 'parent', 'factor', 'offset') - A dictionary of joint_names to the joints that they mimic; compare to the `<mimic>` tag in URDF.  A joint listed here will mimic the movements of the 'parent' joint, subject to the 'factor' and 'offset' provided.  The 'parent' name must be provided, while the 'factor' and 'offset' parameters are optional ( ...
+* `zeros` (dictionary of string -> float) - A dictionary of joint_names to initial starting values for the joint.  In Eloquent, pass on the command-line as '-p zeros.<joint_name>:=<value>'.  Defaults to an empty dictionary, in which case 0 is assumed as the zero for all joints.
```
`<value>` disappears when rendered by github

clalancette replied:
Arg.  Removed the problematic brackets in 2917803.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -139,38 +139,116 @@ def init_urdf(self, robot):
-        self.pub_def_efforts = get_param("publish_default_efforts", False)
-
+    def configure_robot(self, description):
+        self.node.get_logger().info('Got description, configuring robot')
```
this feels more like `debug` than `info` to me

clalancette replied:
Sure.  Fixed in 2917803.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -139,38 +139,116 @@ def init_urdf(self, robot):
+        # a map of name -> dict['parent': parent, factor: factor, offset: offset],
+        # where both factor and offset are optional.  Thus we parse the values we
+        # got into that structure.
+        for name,param in dependent_joints.items():
```
nit space after comma `name, param`. Rock-paper-scissors to decide who makes a follow up PR adding a flake8 test?

clalancette replied:
Haha.  I avoided it here, but yeah, we can do a follow-up to add flake8.  Anyway, style fixed in 2917803.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -139,38 +139,116 @@ def init_urdf(self, robot):
+            newkey = split[0]
+            newvalue = split[1]
+            if newvalue not in ['parent', 'factor', 'offset']:
+                raise Exception("Invalid dependent_joint name '%s'" % (newvalue))
```
Recommend adding `allowed values are ...`

clalancette replied:
Fixed in d2be54f

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -139,38 +139,116 @@ def init_urdf(self, robot):
+            if newkey in dj:
+                dj[newkey].update({newvalue: param.value})
+            else:
+                dj.update({newkey: {newvalue: param.value}})
```
I'm more used to seeing `dj[newkey] = {newvalue: param.value}` for the initial value

clalancette replied:
Fixed in 2917803.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -269,19 +349,26 @@ def loop(self):
+
+            elapsed = time.time() - start_time
+
+            # TODO(clalancette): Use rclpy.Rate once it is available
```
[Rate is available](https://github.com/ros2/rclpy/blob/3b323b564f307cd84d9cb2640c3de8b42ed9295f/rclpy/rclpy/timer.py#L89), but I would definitely recommend a timer callback instead of rate + looping + spin. That would be required to enable this node to be composed.

clalancette replied:
Yeah, excellent call.  Fixed in 2917803.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -269,19 +349,26 @@ def loop(self):
-                pass
+
+            # Spin to make sure we service any subscriptions
+            rclpy.spin_once(self.node, timeout_sec=0)
```
This only executes a single callback. If multiple things are ready, then acting on that data will be delayed by the sleep below. I strongly recommend using a timer instead of looping.

clalancette replied:
Fixed in 2917803.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -299,3 +386,28 @@ def update(self, delta):
                     joint['forward'] = not forward
+
+
+def main(input_args=None):
```
AFAIK, the `console_scripts` entry point doesn't call the function with any arguments, so `input_args` can be deleted.

https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point

clalancette replied:
All right.  The input_args thing is a pattern I've seen elsewhere in the codebase, but I don't feel strongly about it.  Fixed in 2917803.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -139,38 +139,116 @@ def init_urdf(self, robot):
+            self.node.get_logger().info('Waiting for robot_description to be published on the robot_description topic...')
+            self.node.create_subscription(std_msgs.msg.String, 'robot_description',
+                                          lambda msg: self.configure_robot(msg.data),
+                                          rclpy.qos.QoSProfile(depth=10, durability=rclpy.qos.QoSDurabilityPolicy.RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL))
```
nit depth of 1 seems fine. If multiple `robot_description` updates come in a row, I don't think anything is lost as long as the very last one gets acted on.

clalancette replied:
Fixed in 2917803.

clalancette committed:
Updates from review.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
One more review fix.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
@sloretz Thanks for the great review!  Things are now updated as you recommend in almost all cases.  Please take another look when you get a chance.

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -1,32 +1,29 @@
 # Joint State Publisher
 
 This contains a package for publishing `sensor_msgs/JointState` messages for a robot described with URDF.
```
nit, in ROS 2 the CLI tools would say `sensor_msgs/msg/JointState`

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -1,32 +1,29 @@
 
 Published Topics
 ----------------
 * `/joint_states` (`sensor_msgs/JointState`) - The state of all of the movable joints in the system.
```
`sensor_msgs/msg/JointState`

clalancette replied:
Fixed in Fixed in 2832411.

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -1,32 +1,29 @@
 
 Subscribed Topics
 -----------------
+* (optional) `/robot_description` (`std_msgs/String`) - If no URDF is given on the command-line, then this node will listen on the `/robot_description` topic for the URDF to be published.  Once it has been received at least once, this node will start to publish joint values to `/joint_states`.
```
`std_msgs/msg/String`

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -1,32 +1,29 @@
 Subscribed Topics
 -----------------
+* (optional) `/robot_description` (`std_msgs/String`) - If no URDF is given on the command-line, then this node will listen on the `/robot_description` topic for the URDF to be published.  Once it has been received at least once, this node will start to publish joint values to `/joint_states`.
 * (optional) `/any_topic` (`sensor_msgs/JointState`) - If the `sources_list` parameter is not empty (see Parameters below), then every named topic in this parameter will be subscribed to for joint state updates.  Do *not* add the default `/joint_states` topic to this list, as it will end up in an endless loop!
```
`sensor_msgs/msg/JointState`

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -1,32 +1,29 @@
 * `publish_default_efforts` (bool) - Whether to publish a default effort for each movable joint to the `/joint_states` topic.  Defaults to False.
 * `use_mimic_tags` (bool) - Whether to honor `<mimic>` tags in the URDF.  Defaults to True.
 * `use_smallest_joint_limits` (bool) - Whether to honor `<safety_controller>` tags in the URDF.  Defaults to True.
 * `source_list` (array of strings) - Each string in this array represents a topic name.  For each string, create a subscription to the named topic of type `sensor_msgs/JointStates`.  Publication to that topic will update the joints named in the message.  Defaults to an empty array.
```
`sensor_msgs/msg/JointState`

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -1,32 +1,29 @@
-* `dependent_joints` (dictionary of string -> dictionary of 'parent', 'factor', 'offset') - A dictionary of joint_names to the joints that they mimic; compare to the `<mimic>` tag in URDF.  A joint listed here will mimic the movements of the 'parent' joint, subject to the 'factor' and 'offset' provided.  The 'parent' name must be provided, while the 'factor' and 'offset' parameters are optional ( ...
+* `delta` (double) - How much to automatically move joints during each iteration.  Defaults to 0.0.
+* `zeros` (dictionary of string -> float) - A dictionary of joint_names to initial starting values for the joint.  In Eloquent, pass on the command-line as '-p zeros.joint_name:=value'.  Defaults to an empty dictionary, in which case 0 is assumed as the zero for all joints.
+* `dependent_joints` (dictionary of string -> dictionary of 'parent', 'factor', 'offset') - A dictionary of joint_names to the joints that they mimic; compare to the `<mimic>` tag in URDF.  A joint listed here will mimic the movements of the 'parent' joint, subject to the 'factor' and 'offset' provided.  The 'parent' name must be provided, while the 'factor' and 'offset' parameters are optional ( ...
```
mimiced -> mimicked

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher/test/test_copyright.py:
```diff
@@ -0,0 +1,23 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
nit, 2020

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher/test/test_xmllint.py:
```diff
@@ -0,0 +1,23 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
nit 2020

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -0,0 +1,399 @@
+
+def main():
+    # Initialize rclpy with the command-line arguments
+    rclpy.init(args=sys.argv)
```
No need to set `sys.argv`, [rclpy does that already](https://github.com/ros2/rclpy/blob/3b323b564f307cd84d9cb2640c3de8b42ed9295f/rclpy/rclpy/__init__.py#L73)

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py:
```diff
@@ -209,20 +219,53 @@ def randomize_event(self, event):
+            rclpy.spin_once(self.jsp, timeout_sec=0.1)
+
+
+def main(input_args=None):
```
No need for input_args if the `console_scripts` entry point doesn't pass any arguments.

clalancette replied:
Fixed in 2832411.

sloretz commented on joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py:
```diff
@@ -209,20 +219,53 @@ def randomize_event(self, event):
+
+    app = QApplication(sys.argv)
+    jsp_gui = JointStatePublisherGui('Joint State Publisher',
+                                     JointStatePublisher(parsed_args.urdf_file))
```
I launched the GUI with no urdf via CLI, and then published it via `/robot_description`, but the sliders don't show up. Is something supposed to update the GUI when that happens?

clalancette replied:
Yep, this was another missing feature.  Thanks for catching.  Fixed in 2832411.

sloretz commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -0,0 +1,399 @@
+                joint = self.free_joints[parent]
+
+            if has_position and 'position' in joint:
+                msg.position[i] = float(joint['position']) * factor + offset
```
I get an `IndexError` here when I publish the same URDF twice on the `/robot_description` topic. It looks like `self.joint_list` has duplicate joint names, but `num_joints` was calculated from `free_joints` and `dependent_joints` which can't have duplicates because they're dictionaries, so `msg.position` is half the length. Maybe the variables that are set in `__init__` need to be set in `self.configure_robot` instead so they get cleared when a new description is published?

clalancette replied:
This was absolutely a problem.  Fixing it turned out to be...complicated.  Commit 735f2a7 successfully fixes it, along with making the layout way more dynamic and useful (in my opinion).  It also simplifies the code a bit, at the expense of adding a new PyQt layout type (FlowLayout), and putting the Center/Randomize buttons at the top.  I think that is a pretty good tradeoff for a dynamic, reflowable UI, but please take a look and let me know what you think.

clalancette committed:
More fixes from review.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Allow dynamic updates in the GUI

This means that updates from the robot_description topic are
properly shown in the UI now.  While I was in here, I also
changed it so that the UI "flows" as you resize the window.
This should be much more usable on smaller screens.  It
also simplifies the code quite a bit.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
@sloretz All of your above points should be fixed now.  Ready for another round, thanks!

sloretz committed:
Initialize gui sliders at startup

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add 64_joint_robot launch test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Delete ROS 1 64 joint rostest

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix exception about too many arguments

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Delete ROS 1 mimic cycle rostest

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add mimic cycle launch test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add offset and multiplier attributes

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add launch test for mimic chain

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Delete ROS 1 mimic chain rostest

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix collada joints ignoring publish_default_*

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Delete ROS 1 collada rostest

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add test comparing urdf and collada joints

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove ROS 1 urdf rostest

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add zero joint launch test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Delete ROS 1 zero joint rostest

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

clalancette committed:
Remove trailing whitespace.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Fix up a comment.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

sloretz committed:
Test URDF/DAE from ROS topic

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>
